vortex-layout/src/layouts/flat/evaluator.rs

```diff
@@ -1,33 +1,55 @@
 use async_trait::async_trait;
+use flatbuffers::root;
+use futures::future::try_join_all;
 use vortex_array::compute::filter;
+use vortex_array::parts::ArrayParts;
 use vortex_array::ArrayData;
-use vortex_error::{vortex_bail, vortex_err, VortexResult};
+use vortex_error::{vortex_err, VortexExpect, VortexResult};
 use vortex_expr::ExprRef;
-use vortex_ipc::messages::{BufMessageReader, DecoderMessage};
+use vortex_flatbuffers::{array as fba, FlatBuffer};
 use vortex_scan::{AsyncEvaluator, RowMask};
 
 use crate::layouts::flat::reader::FlatReader;
 use crate::reader::LayoutScanExt;
+use crate::LayoutReader;
 
 #[async_trait(?Send)]
 impl AsyncEvaluator for FlatReader {
     async fn evaluate(self: &Self, row_mask: RowMask, expr: ExprRef) -> VortexResult<ArrayData> {
-        // Grab the byte buffer for the segment.
-        let bytes = self.segments().get(self.segment_id()).await?;
-
-        // Decode the ArrayParts from the message bytes.
-        // TODO(ngates): ArrayParts should probably live in vortex-array, and not required
-        //  IPC message framing to read or write.
-        let mut msg_reader = BufMessageReader::new(bytes);
-        let array = if let DecoderMessage::Array(parts) = msg_reader
-            .next()
-            .ok_or_else(|| vortex_err!("Flat message body missing"))??
-        {
-            parts.decode(self.ctx(), self.dtype().clone())
-        } else {
-            vortex_bail!("Flat message is not ArrayParts")
-        }?;
+        // Fetch all the array buffers.
+        let buffers = try_join_all(
+            (0..self.layout().nsegments())
+                .map(|idx| {
+                    self.layout()
+                        .segment_id(idx)
+                        .vortex_expect("bounds checked")
+                })
+                .map(|segment_id| self.segments().get(segment_id)),
+        )
+        .await?;
 
+        // Fetch the array flatbuffer.
+        let flatbuffer = FlatBuffer::try_from(
+            buffers
+                .last()
+                .ok_or_else(|| vortex_err!("Flat message missing"))?
+                .clone(),
+        )?;
+
+        let row_count = usize::try_from(self.layout().row_count())
+            .vortex_expect("FlatLayout row count does not fit within usize");
+
+        let array_parts = ArrayParts::new(
+            row_count,
+            root::<fba::Array>(flatbuffer.as_ref()).vortex_expect("Invalid fba::Array flatbuffer"),
+            flatbuffer.clone(),
+            buffers.iter().take(buffers.len() - 1).cloned().collect(),
```

Comment:
This seems silly. You can pop the last element and then pass the buffers here directly